File: nodejs/backend/src/controller/userController.ts
import { Request, Response } from 'express';
import { UserService } from '../service/UserService';
import { toUserResponseDto } from '../dto/authDto';
import { requireStringParam } from '../error/paramHelpers';

export class UserController {
  constructor(private readonly userService: UserService) {}

  list = async (_req: Request, res: Response): Promise<void> => {
    const users = await this.userService.list();
    res.json(users.map(toUserResponseDto));
  };

  remove = async (req: Request, res: Response): Promise<void> => {
    await this.userService.delete(requireStringParam(req, 'id'));
    res.status(204).send();
  };
}

File: rust/backend/src/controller/user_controller.rs
use crate::{dto::UserResponse, error::AppError, state::AppState};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use uuid::Uuid;

pub async fn list_users(
    State(state): State<AppState>,
) -> Result<impl IntoResponse, AppError> {
    let users: Vec<UserResponse> = state
        .user_service
        .list()
        .await?
        .into_iter()
        .map(UserResponse::from)
        .collect();

    Ok(Json(users))
}

pub async fn delete_user(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    state.user_service.delete(id).await?;
    Ok(StatusCode::NO_CONTENT)
}
